### ISSUE TITLE
Support iPadOS (iPad) & iOS (iPhone) apps

### ISSUE THREAD

**Issue by dbrookman** [mas-cli/mas#321]
# Notes

It appears that the data structure in `CKSoftwareMap` (from CommerceKit, a macOS private framework) that lists the apps installed from the Mac App Store (which is used by mas) doesn't include installed iOS or iPadOS apps.

Spotlight contains some info about installed iOS & iPadOS along with installed macOS apps, but it does not seem to include the ADAM IDs for iOS or iPadOS apps.

We can look for other data sources that contain iOS & iPadOS along with their ADAM IDs (investigating private frameworks depends on #565).

Assuming that there are no other useful data sources, and that the known data sources behave as I suspect from the info I've seen, mas can:

## Find installed apps via Spotlight instead of via `CKSoftwareMap` (done in mas 2.0.0)
 
- ADAM IDs:
  - While Spotlight includes some data about iOS & iPadOS apps, it doesn't include ADAM IDs for them (but it does include ADAM IDs for macOS apps from the Mac App Store)
  - Spotlight does, however, include bundle IDs, which are unique per app.
  - Implement #541
  - Encourage users to use bundle IDs instead of ADAM IDs
  - If users use ADAM IDs, use the Apple iTunes Search Web API to lookup bundle IDs from ADAM IDs
- App names:
  - Spotlight includes all the data that mas currently reads from objects returned from `CKSoftwareMap`, except it doesn't contain app names; it contains some data that might be similar to app names, but that can be off.
  - To provide missing app names, mas can query the Apple iTunes Search Web API.

The above is ridiculous, but it is the only currently known generic workaround.

## Investigate for better sources of installed app data

- `pkgutil --pkg-info-plist <bundle-id>`
- `/Applications/<app-dir-name>.app/Contents/_MASReceipt/receipt`

See:

- #648
- #669

# Original Issue

## Your Environment

- mas version: 1.8.0
- macOS version: 11.2.1 (20D74)

### mas Install Method

`brew install mas` (homebrew-core)

## Describe the Bug

`mas list`, `mas search`, `mas outdated` do not show iOS apps that can be installed via the App Store on M1 macs.

## To Reproduce

1. Install an iOS app via the App Store.
2. Run `mas list`

## Expected Behavior

The installed iOS app should be among the apps listed.

## Actual Behavior

The iOS app is not among the apps listed.

**Comment by phatblat** [comment]
The is a great idea for a new feature. I had to send my DTK back, but while I had it I only installed iOS app through Xcode. Can you install them from the Mac App Store?

**Comment by dbrookman** [comment]
Yeah, you install them from the mac App Store.
When you search you're presented with the option to see iOS apps that you can install.
<img width="221" alt="Screen Shot 2021-03-21 at 11 22 57 AM" src="https://user-images.githubusercontent.com/53625739/111910452-cbaed880-8a37-11eb-8717-0b46009cbb71.png">

**Comment by phatblat** [comment]
One thing that will need to change to support this is to change how the search and lookup commands work. Today they use the iTunes search web API and that has a filter for only macOS. Once we know the app identifier it may "just work".

One thing you could try is to copy the URL for an iOS app from the store and look for the numeric identifier at the end (drop any characters). Then try `mas install 1234` with that ID.

**Comment by dbrookman** [comment]
You're right, you can get the ID from the App Store website URL.
I specifically tried [Desert Golfing](https://apps.apple.com/us/app/desert-golfing/id902062673) (ID: 902062673)

`mas info 902062673` works, returning:
```
Desert Golfing 1.22 [1.99]
By: Captain Games Inc.
Released: 2017-12-05
Minimum OS: 9.0
Size: 2.2 MB
From: https://apps.apple.com/us/app/desert-golfing/id902062673?uo=4
```

`mas install 902062673` brings up the following dialog:
<img width="372" alt="Screen Shot 2021-03-22 at 10 55 41 PM" src="https://user-images.githubusercontent.com/53625739/112085960-ce0d5180-8b61-11eb-88b5-f9373c8f624d.png">
I suspect the Minimum OS mismatch might be the issue here.
Clicking "Download Last Compatible" returns:
```
==> Downloading Desert Golfing
Error: Download failed: The installation could not be started.
```
Clicking "Cancel" (rather obviously) returns:
```
Error: Download request failed: User selected cancel button
```

Just to be sure, I double checked, and it installs from the App Store fine.

**Comment by iMonZ** [comment]
Would be so excited if this would be fixed!!!

**Comment by natterstefan** [comment]
Me too!

**Comment by jeeftor** [comment]
Is this being worked on or nay?

**Comment by chkpwd** [comment]
Pinging here to show support as well.

**Comment by aescolastico** [comment]
+1 to hoping this gets support

**Comment by rgoldberg** [comment]
I doubt this will be worked on until summer 2025, when the M4 Mac Studios are expected to be released. Until then, I'm stuck on Intel.

It can be done sooner if someone else wants to work on it, or if someone wants to give me a free Apple Silicon Mac :).

**Comment by berndkrannich** [comment]
Fwiw, I'm using [majd/ipatool](https://github.com/majd/ipatool) for downloading IPA app packages for iOS. Not sure if the mechanism is the same as for Mac apps and/or if there's potential for reuse/collaboration between the projects.

**Comment by andrewcrook** [comment]
+1 for this

**Comment by AlbertoPimpo** [comment]
Any hope to see this soon?

**Comment by rgoldberg** [comment]
Unfortunately, no.

Besides the problems I prepended to the original issue description, the Apple private framework functions mas uses to install macOS apps don't work for iOS or iPadOS apps. I forget what the exact problem is.

I imagine that newer Apple private framework functions support installing iOS / iPadOS apps, but, whenever I've tried to use the likely current functions from AppStoreDaemon, I got entitlement errors. Attempting to use certain likely entitlements caused the app to crash, so it looks like they're only usable by code signed by Apple.

If anyone else wants to investigate this, they can, but I don't want to continually run into seemingly insurmountable problems.

**Comment by brian163** [comment]
> I imagine that newer Apple private framework functions support installing iOS / iPadOS apps, but, whenever I've tried to use the likely current functions from AppStoreDaemon, I got entitlement errors. Attempting to use certain likely entitlements caused the app to crash, so it looks like they're only usable by code signed by Apple.

Might a similar explanation apply to apps badged as "Apple Arcade" in the App Store? As an example, mas list includes the ADAM ID for this Apple Arcade app but mas outdated does not show the 1.08.01 update showing in the App Store:

% mas list
6545262709  Pro Tour Golf                         (1.08.00)

% mas outdated
(no entries)

%mas info 6545262709
Error: No apps found in the App Store for ADAM ID 6545262709

(If this is an unrelated issue I can create a separate ticket for it.)

**Comment by rgoldberg** [comment]
@brian163 Thanks for the info.

`info` & `outdated` correctly don't find `Pro Tour Golf` because it's an iOS / iPadOS app, not a macOS app.

`list` shouldn't find it, either, but it appears that Spotlight has a `kMDItemAppStoreAdamID` value for `Pro Tour Golf`, unlike how Spotlight doesn't have it for the other iOS / iPadOS that I've tried. That might be the case for all Apple Arcade games, or for just certain select ones.

Can you run the 2 following commands from the terminal to investigate?

```shell
mdfind 'kMDItemAppStoreAdamID = 6545262709' 
mdls "$(mdfind 'kMDItemAppStoreAdamID = 6545262709')"
```

**Comment by brian163** [comment]
> `info` & `outdated` correctly don't find `Pro Tour Golf` because it's an iOS / iPadOS app, not a macOS app.
> 
So unlike some iOS/iPadOS apps listed in the (macOS) App Store as such, Pro Tour Golf is actually listed as "Mac, iPad, iPhone, Apple TV". 

<img width="1004" height="163" alt="Image" src="https://github.com/user-attachments/assets/ff9ba073-c5db-485b-92bb-d804888ea22a" />

App Store link, if helpful:
https://apps.apple.com/us/app/pga-tour-pro-golf/id6545262709

> `list` shouldn't find it, either, but it appears that Spotlight has a `kMDItemAppStoreAdamID` value for `Pro Tour Golf`, unlike how Spotlight doesn't have it for the other iOS / iPadOS that I've tried. That might be the case for all Apple Arcade games, or for just certain select ones.
> 
> Can you run the 2 following commands from the terminal to investigate?
> 
> mdfind 'kMDItemAppStoreAdamID = 6545262709' 
> mdls "$(mdfind 'kMDItemAppStoreAdamID = 6545262709')"

**% mdfind 'kMDItemAppStoreAdamID = 6545262709'** 
/Applications/Pro Tour Golf.app

**mdls "$(mdfind 'kMDItemAppStoreAdamID = 6545262709')"**

```
_kMDItemDisplayNameWithExtensions       = "Pro Tour Golf.app"
kMDItemAlternateNames                   = (
    "Pro Tour Golf.app"
)
kMDItemAppStoreAdamID                   = 6545262709
kMDItemAppStoreCategory                 = "Games"
kMDItemAppStoreCategoryType             = "public.app-category.games"
kMDItemAppStoreHasReceipt               = 1
kMDItemAppStoreInstallerVersionID       = "881296324"
kMDItemAppStoreIsAppleSigned            = 1
kMDItemAppStoreParentalControls         = "4+"
kMDItemAppStorePurchaseDate             = 2026-01-21 03:08:52 +0000
kMDItemAppStoreReceiptIsMachineLicensed = 0
kMDItemAppStoreReceiptIsRevoked         = 0
kMDItemAppStoreReceiptIsVPPLicensed     = 0
kMDItemAppStoreReceiptType              = "Production"
kMDItemCFBundleIdentifier               = "com.hypgames.protourgolf"
kMDItemContentCreationDate              = 2026-01-15 19:48:10 +0000
kMDItemContentCreationDate_Ranking      = 2026-01-15 00:00:00 +0000
kMDItemContentModificationDate          = 2026-01-21 03:08:53 +0000
kMDItemContentType                      = "com.apple.application-bundle"
kMDItemContentTypeTree                  = (
    "com.apple.application-bundle",
    "com.apple.application",
    "public.executable",
    "com.apple.localizable-name-bundle",
    "com.apple.bundle",
    "public.directory",
    "public.item",
    "com.apple.package"
)
kMDItemCopyright                        = ""
kMDItemDateAdded                        = 2026-02-11 21:14:03 +0000
kMDItemDescription                      = ""
kMDItemDisplayName                      = "Pro Tour Golf"
kMDItemDocumentIdentifier               = 0
kMDItemExecutableArchitectures          = (
    arm64,
    "x86_64"
)
kMDItemFSContentChangeDate              = 2026-01-21 03:08:53 +0000
kMDItemFSCreationDate                   = 2026-01-15 19:48:10 +0000
kMDItemFSCreatorCode                    = ""
kMDItemFSFinderFlags                    = 0
kMDItemFSHasCustomIcon                  = (null)
kMDItemFSInvisible                      = 0
kMDItemFSIsExtensionHidden              = 1
kMDItemFSIsStationery                   = (null)
kMDItemFSLabel                          = 0
kMDItemFSName                           = "Pro Tour Golf.app"
kMDItemFSNodeCount                      = 1
kMDItemFSOwnerGroupID                   = 0
kMDItemFSOwnerUserID                    = 0
kMDItemFSSize                           = 4104288874
kMDItemFSTypeCode                       = ""
kMDItemInterestingDate_Ranking          = 2026-01-21 00:00:00 +0000
kMDItemKeywords                         = ""
kMDItemKind                             = "Application"
kMDItemLogicalSize                      = 4104288874
kMDItemPhysicalSize                     = 4097466368
kMDItemVersion                          = "1.08.00"
```
Thanks for looking into this.

**Comment by rgoldberg** [comment]
Thanks for the info.

The iTunes Search web API does not recognize PGA TOUR Pro Golf as desktop software, which causes `info` & `outdated` to ignore it. The API does, however, indicate that the game supports the `MacDesktop-MacDesktop` "device", which is something that I haven't seen before.

For various reasons, it seems like PGA TOUR Pro Golf actually has a macOS version (distinct from an iOS / iPadOS version that can run on macOS), despite not being categorized as a desktop app.

mas doesn't currently support iOS-only / iPadOS-only apps that can still be run on macOS, because the ones I have seen all behave differently than macOS apps, including the Apple install/update app function not working for them, the iTunes Search web API not considering them as desktop apps, Spotlight not containing ADAM IDs for them, etc..

If the Apple install/update app function works for apps like PGA TOUR Pro Golf, then I can look into supporting it and other apps like it (presumably all `MacDesktop-MacDesktop` "device" apps). It might get tricky though, for various reasons with which I don't want to inundate you right now.

Can you try reinstalling PGA TOUR Pro Golf via the following mas command to let me know if the Apple install/update app function works for it?

```shell
mas install --force 6545262709
```

**Comment by brian163** [comment]
I will give this a try first thing tomorrow.

**Comment by brian163** [comment]
I realized this probably wasn't going to work since 'mas info' had previously returned "Error: No apps found in the App Store for ADAM ID 6545262709".

But I tried just to be sure and did receive the same error. 

Apple clearly has some black magic fluckery going on with Apple Arcade (the "Games" app) and some tie-in to the App Store. I appreciate you looking at it for me though.

**Comment by rgoldberg** [comment]
Thanks for trying that.

I have a local branch that I assume should fix the problem. I've verified that it fixes `info`/`lookup`, but I cannot test `install` for PGA TOUR Pro Golf because I do not have Apple Arcade.

If I make a PR, can you build the PR from source & test `mas install --force 6545262709` again? Or would it be much easier for you to test it after it's been merged into main? The former is better so that I don't publish code that might not work 100% (it should work, but better to test before merging, if possible).

The one limitation is that, while apps like PGA TOUR Pro Golf will presumably work for most mas commands, `search` & `lucky` will not find apps like PGA TOUR Pro Golf. I have an idea about how to get those 2 commands to support these apps, but it's ugly, so I'd prefer to just fix all the other commands first.

**Comment by rgoldberg** [comment]
Can you open a new issue with the relevant info and a link to your first comment here?

The issue should be entitled "Support `MacDesktop-MacDesktop` apps". (the double repetition of `MacDesktop` is correct)

**Comment by TyceHerrman** [comment]
what about a smaller-scope feature? A separate command group someting like `mas ios-on-mac`.

wouldn't do full install/update support, but could do:

- list: use Spotlight metadata to report bundle ID, path, and local version when an iOS/iPadOS app lacks an ADAM ID.
- check <bundle-id>: public iTunes Lookup comparison; explicitly advisory.
- open <bundle-id>: open the App Store product page for user action.

This would at least provide inventory, drift reporting, and Store handoff for ios/ipados apps on mac.

**Comment by rgoldberg** [comment]
It's possible, but I've been working on many other larger issues first (see list at end), some of which would help enable implementing parts your request, or would help users more easily use parts of the new feature.

If partial support were to be added, it would not involve new commands.

When app IDs are (or `--all` is) supplied to a command for which non-macOS apps can be supported, the command should process each app ID regardless of the app's platform.

When a command is run without app IDs (or `--all`), then the command should support a `--platforms` option:

```ebnf
platforms_option ::= "--platforms" <platform> [ "," <platform> ]…`
platform ::= "macos" | "ios" | "ipados" | "all"
```

`<platform>` would probably be case insensitive.

`--platforms` would default to `macos` for `search`, and either `macos` or `all` for `list` & `outdated`.

Once the configuration system (#838) has been released, users could set their own default platform set, and thus avoid typing it in every command line.

## Commands

Relevant commands accepting app IDs (or `--all`) to select apps should process apps for any platform when such arguments are supplied. Such commands include:

- `home`
- `list`
- `lookup`/`info`
- `open`
- `outdated` (mutually exclusive with `--accurate`)
- `seller`/`vendor`
- `uninstall`

Relevant commands that process all apps installed apps when no app ID arguments are supplied should support `--platforms`. Should the default be `all` or `macos`? `all` would show more info than would `macos`; users would, however, probably expect to be able to update any listed apps, but wouldn't be able to update any non-macOS ones. If we added a platform output field, then maybe `all` would be OK. Also, if we output a `<platform> apps not supported` error / warning for `<platform>` != `macos` in commands like `install`, `update`, etc., that might alleviate user confusion. Such commands include:

- `list`
- `outdated` (mutually exclusive with `--accurate`)

Relevant commands that do not accept app ID arguments (or `--all`) should similarly support `--platforms`, but I'd be even more likely to support `macos` instead of `all` as the default. Even if a platform output field & improved non-macOS errors / warn gins were implemented, including non-macOS apps in search results could crowd out macOS app search results. Such commands include:

- `search`

I'll see if I can fit this in for 7.1.0. I might implement this first for only when app IDs (or `--all`) are supplied to relevant commands, not when commands are run without app IDs (or `--all`).

## Current & planned work

- Ready for release in 7.1.0 (local, not yet in GitHub):
  - massive concurrency overhaul
  - massive subprocess & privileges overhaul, including #1193
  - code base cleanup
  - update to Swift 6.3 & current dependencies
- Pending responses from third parties before implementation for 7.1.0:
  - #1248
  - #1266
  - #1274
- Ready for late 2026:
  - #1024
- Implementing for later mas versions (first 2 of which require extensive updates to third-party libraries):
  - #838
  - #678
  - #1226
- Haven't started, but plan to work on many issues dependent on (or vastly improved by) #838, like:
  - #1148
  - #386
  - #75
  - #839

**Comment by rgoldberg** [comment]
@TyceHerrman Please subscribe to #1275.
